File: Sources/OpenSwiftUIShims/LockedPointer.m
//
//  LockedPointer.c
//  
//
//  Created by Kyle on 2023/10/19.
//

#import "LockedPointer.h"

LockedPointer _LockedPointerCreate(size_t size) {
    LockedPointer ptr = malloc(size);
    if (ptr == NULL) { abort(); }
    ptr->lock = OS_UNFAIR_LOCK_INIT;
    ptr->offset = 8;
    return ptr;
}

void *_LockedPointerGetAddress(LockedPointer ptr) {
    return (void *)((uintptr_t)ptr + (uintptr_t)(ptr->offset));
}

void _LockedPointerDelete(LockedPointer ptr) {
    free(ptr);
}

void _LockedPointerUnlock(LockedPointer ptr) {
    os_unfair_lock_unlock(&ptr->lock);
}

void _LockedPointerLock(LockedPointer ptr) {
    os_unfair_lock_lock(&ptr->lock);
}

File: Sources/OpenSwiftUI/Shims/LockedPointer.swift
//
//  LockedPointer.swift
//  OpenSwiftUI
//
//  Created by Kyle on 2023/10/18.
//  Lastest Version: iOS 15.5
//  Status: Complete

internal import OpenSwiftUIShims

extension LockedPointer {
    @_transparent
    init<Data>(type: Data.Type) {
        self = _LockedPointerCreate(MemoryLayout<Data>.size, MemoryLayout<Data>.alignment)
    }

    @_transparent
    @discardableResult
    func withUnsafeMutablePointer<Data, R>(_ type: Data.Type = Data.self, _ body: (UnsafeMutablePointer<Data>) throws -> R) rethrows -> R {
        try body(withUnsafeMutablePointer(type))
    }

    @_transparent
    func withUnsafeMutablePointer<Data>(_ type: Data.Type = Data.self) -> UnsafeMutablePointer<Data> {
        _LockedPointerGetAddress(self).assumingMemoryBound(to: Data.self)
    }

    @_transparent
    func lock() {
        _LockedPointerLock(self)
    }

    @_transparent
    func unlock() {
        _LockedPointerUnlock(self)
    }

    @_transparent
    func delete() {
        _LockedPointerDelete(self)
    }
}
